examples/Example7/example7-adept.cc

```diff
@@ -0,0 +1,201 @@
+}
+
+int main(int argc, char* argv[])
+{
+#ifndef VECGEOM_USE_NAVINDEX
+  std::cout << "### VecGeom must be compiled with USE_NAVINDEX support to run this.\n";
+  return 1;
+#endif
+
+  // Command line options
+  // G4HepEMData file path is stored as an auxiliary
+  OPTION_STRING(gdml_file, "");
+
+  // Remainder of options follow TestEm3 example
+  OPTION_INT(cache_depth, 0); // 0 = full depth
+  OPTION_INT(particles, 1000);
+  OPTION_DOUBLE(energy, 10); // entered in GeV
+  energy *= copcore::units::GeV;
+  OPTION_INT(batch, -1);
+
+  // File arg must not be empty
+  if(gdml_file.empty()) {
+    std::cerr << "required gdml_file argument is empty\n";
+    return 2;
+  }
+  // Main task in this example is to reconstitute GDML and G4HepEMData
+  // - GDML
+  // Inputs: File path, Cache depth
+  // Output: vgdml::Middleware, vecgeom::VPlacedVolume (world volume)
+  vecgeom::GeoManager::Instance().SetTransformationCacheDepth(cache_depth);
+  vgdml::Parser vgdmlParser;
+  auto middleWare = vgdmlParser.Load(gdml_file.c_str(), false, copcore::units::mm);
+  if (middleWare == nullptr)
+  {
+    std::cerr << "Failed to read geometry from GDML file '" << gdml_file << "'" << std::endl;
+    return 3;
+  }
+
+  const vecgeom::VPlacedVolume *world = vecgeom::GeoManager::Instance().GetWorld();
+  if (world == nullptr) {
+    std::cerr << "GeoManager world volume is nullptr" << std::endl;
+    return 3;
+  }
+#ifdef VERBOSE
+  std::cout << "World (ID " << world->id() << ")" << std::endl;
+  PrintDaughters(world);
+#endif
+
+  // G4HepEmData
+  // Input: file path (here from vgdml::Middleware)
+  // Output: G4HepEmState pointer (owned)
+  const auto& userInfo = middleWare->GetUserInfo();
+  vgdml::Auxiliary g4hepemFileAux = find_by_type(userInfo, "g4hepemfile");
+
+  if (g4hepemFileAux == vgdml::Auxiliary{}) {
+    std::cerr << "GDML file '" << gdml_file << "' has no 'g4hepemfile' user info auxiliary" << std::endl;
+    return 4;
+  }
+
+  std::string g4hepem_file = g4hepemFileAux.GetValue();
+  std::ifstream jsonIS{ g4hepem_file.c_str() };
+
+  G4HepEmState* inState = G4HepEmStateFromJson(jsonIS);
+  if(inState == nullptr)
+  {
+    std::cerr << "Failed to read G4HepEmState from " << g4hepem_file << std::endl;
+    return 4;
+  }
+
+  // Connect up VecGeom _physical_ volume ids to G4HepEMData material cut couple indexes
+  // via Geant4 material cut couple index supplied as a volume auxiliary in the GDML
+  // Inputs: vecgeom::GeoManager, vgdml::Middleware, G4HepEmData
+  // Output: Array of ints. Array indices are VecGeom PV id numbers, value at that index is G4HepEmData MCC index
+  // Visting all PVs walks the tree, so we get multiple copies in the vector. Thus reduced to a
+  // set of pointers afterwards to uniq the list.
+  std::vector<vecgeom::VPlacedVolume*> tmpPlacedVolumes;
+  vecgeom::GeoManager::Instance().getAllPlacedVolumes(tmpPlacedVolumes);
+  std::set<vecgeom::VPlacedVolume*> allPlacedVolumes{tmpPlacedVolumes.begin(), tmpPlacedVolumes.end()};
```

Comment:
`std::unordered_set` probably suffices, right? It doesn't make sense to sort pointers...